Refactor ReshapeSparseTensor into a template+class

benbarsdell opened the pull request:
This is in preparation for adding a GPU implementation.
No functional change.

cc @nluehr

benbarsdell committed:
Refactor ReshapeSparseTensor into a template+class

- This is in preparation for adding a GPU implementation.
- No functional change.

cheshire commented:
@allenl Any idea who should review this?

cheshire commented:
@allenlavoie Any idea who should review this?

allenlavoie commented:
@penpornk has been working on sparse kernels I believe. Penporn, want to take a look?

penpornk commented:
@allenlavoie @cheshire Sure, I can review this. :)

penpornk commented on tensorflow/core/kernels/reshape_util.cc:
```diff
@@ -31,6 +31,54 @@ limitations under the License.
+
+template <>
+struct ReshapeSparseTensor<CPUDevice> {
+  Status operator()(OpKernelContext *context, const TensorShape &input_shape,
```
`context` is not used and should be removed from the parameter list. We can add it back later if you use it in the GPU implementation.

benbarsdell replied:
Done.

penpornk commented on tensorflow/core/kernels/reshape_util.h:
```diff
@@ -16,18 +16,36 @@ limitations under the License.
+namespace functor {
+
+template <typename Device>
+struct ReshapeSparseTensor {
```
Nit: For clarity, let's rename this to `ReshapeSparseTensorFunctor`.

benbarsdell replied:
Done.

penpornk commented on tensorflow/core/kernels/reshape_util.cc:
```diff
@@ -31,6 +31,54 @@ limitations under the License.
+                    typename TTypes<int64>::ConstMatrix input_indices,
+                    typename TTypes<int64>::Matrix output_indices) const {
+    (void)context;
+    int64 input_rank = input_shape.dims();
+    int64 output_rank = output_shape.dims();
+    int64 nnz = input_indices.dimension(0);
```
Nit: Let's make them `const`.

benbarsdell replied:
Done.

penpornk commented on tensorflow/core/kernels/reshape_util.cc:
```diff
@@ -153,6 +167,26 @@ void ReshapeSparseTensor(OpKernelContext *context,
+  }
 }
 
+#define EXPLICITLY_INSTANTIATE_FUNCTION(Device)                    \
+  template void ReshapeSparseTensor<Device>(                       \
+      OpKernelContext *context, const Tensor &input_indices_in,    \
+      const Tensor &input_shape_in, const Tensor &target_shape_in, \
+      int output_indices_idx, int output_shape_idx)
+EXPLICITLY_INSTANTIATE_FUNCTION(CPUDevice);
+#undef EXPLICITLY_INSTANTIATE_FUNCTION
```
Shouldn't this be in the header?

benbarsdell replied:
Explicitly instantiating in the source file means the function will only be linked once instead of  being inlined and instantiated in multiple translation units.

penpornk replied:
Thank you for the explanation!

penpornk reviewed (changes requested):
Thank you for the PR and I'm so sorry for the delay! I have minor comments.

benbarsdell committed:
Address minor review comments for PR 46275

- Remove unused context parameter.
- Rename functor::ReshapeSparseTensor to functor::ReshapeSparseTensorFunctor.
- Make integer constants const.

benbarsdell commented:
Gentle ping.

penpornk reviewed (approved):
I'm sorry again for the delay!
